### PR TITLE
feat(datafile-management): Introducing config manager for managing project config - WIP

### PR DESCRIPTION
**Pull request by rashidsp**
### Summary
This PR introduces config manager which polls for datafile after some interval and prepares project config based on it if it has updated.

### CODE REVIEW

File: lib/optimizely/config_manager/fallback_project_config_manager.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#
+
+require_relative 'project_config_manager'
+
+module Optimizely
+  class FallbackProjectConfigManager < ProjectConfigManager
+    # Implementation of ProjectConfigManager interface.
+    # Returns the stored ProjectConfig instance.
+
+    def initialize(datafile = nil, logger = nil, error_handler = nil, skip_json_validation = false)
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/fallback_project_config_manager.rb]: Fallback projectconfigmanager should only take `ProjectConfig`. to create DatafileProjectConfig from `datafile` must be in any utils method.

File: lib/optimizely/config_manager/fallback_project_config_manager.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    end
+
+    def get_config
+      # Returns object ProjectConfig instance.
+      @config
+    end
+
+    private
+
+    def set_config(datafile)
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/fallback_project_config_manager.rb]: we don't really need it. Make it simple.

File: lib/optimizely/config_manager/polling_project_config_manager.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        )
+        return
+      end
+
+      handle_response response
+    end
+
+    def running
+      # Check if polling thread is alive or not.
+      @polling_thread.alive?
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/polling_project_config_manager.rb]: Where is `auto_update`?

Comment (msohailhussain) [inline lib/optimizely/config_manager/polling_project_config_manager.rb]: where stopping polling_thread

File: lib/optimizely/config_manager/polling_project_config_manager.rb [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    end
+
+    def run
+      # Triggered as part of the thread which fetches the datafile and sleeps until next update interval.
+      while running
+        fetch_datafile
+        sleep @update_interval
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/polling_project_config_manager.rb]: Please add a check if we don't get response before our next schduled call, then no need to fetch again and as soon as get the response, then fetch datafile again immdiately.

File: lib/optimizely/config_manager/polling_config_manager.rb [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      # Helper method to determine URL from where to fetch the datafile.
+      # sdk_key - Key uniquely identifying the datafile.
+      # url - String representing URL from which to fetch the datafile.
+      # url_template - String representing template which is filled in with
+      #               SDK key to determine URL from which to fetch the datafile.
+      # Returns String representing URL to fetch datafile from.
+
+      raise InvalidInputsError, 'Must provide at least one of sdk_key or url.' if sdk_key.nil? && url.nil?
+
+      return (url_template % sdk_key) unless url
```
Comment (oakbani) [inline lib/optimizely/config_manager/polling_config_manager.rb]: can this raise an exception if url_template doesn't have a placeholder or more? or the type is different? If yes, we should catch it

File: lib/optimizely/config_manager/static_config_manager.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      rescue StandardError => e
+        @logger = SimpleLogger.new
+        error_msg = e.class == InvalidDatafileVersionError ? e.message : InvalidInputError.new('datafile').message
+        error_to_handle = e.class == InvalidDatafileVersionError ? InvalidDatafileVersionError : InvalidInputError
+        @logger.log(Logger::ERROR, error_msg)
+        @error_handler.handle_error error_to_handle
+        return
+      end
+
+      previous_revision = @config.revision if @config
```
[the remark this answers is not in the document]
Comment (rashidsp) [reply lib/optimizely/config_manager/static_config_manager.rb]: Polling config manager overrides this method!

File: lib/optimizely/config_manager/polling_config_manager.rb [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+require_relative 'helpers/validator'
+require_relative 'logger'
+require_relative 'notification_center'
+require_relative 'project_config'
+require_relative 'static_config_manager'
+
+require 'net/http'
+
+module Optimizely
+  class PollingConfigManager < StaticConfigManager
```
Comment (mikeng13) [inline lib/optimizely/config_manager/polling_config_manager.rb]: This should not be derived from the `StaticConfigManager`. 

The `StaticConfigManager` should only be a holder for the proejct config and should implement the `BaseConfigManager`

The `PollingConfigManager` is another implementation of the `BaseConfigManager`.

**Comment by msohailhussain** [comment]
> Instead of working on a feature branch, please open up this PR against master!

Yes Mike, it was supposed to be merge in `master`, This PR is actually not finalized yet.

**Comment by mikeng13** [comment]
Closing this as we've decided to split up into separate PRs.

**Pull request closed without merging**